### PR TITLE
Force rotation X.509 SVIDs in Agent side

### LINKED ISSUE #3903: Rotate X.509 Workload SVIDs using tainted keys
Update agent sync to rotate all cached SVIDs affected by a tainted Key

Depends on #3901

### LINKED ISSUE #3907: Force rotation of Agent SVID
Force rotation of Agent SVID when it was signed by a tainted Key

Depends on #3901

### PR DESCRIPTION
**Pull request by MarcosDY**
* Force rotation of X.509 workload SVIDs in lru cache
* Force rotation of X.509 workload SVIDs in store SVID cache
* Force rotation of Agent SVID

**Which issue this PR fixes**
fixes #3907
fixes #3903

### CODE REVIEW

File: pkg/common/x509util/cert.go [hunk 1]
```diff
@@ -72,3 +72,28 @@ func RawCertsFromCertificates(certs []*x509.Certificate) [][]byte {
...
+		intermediatePool.AddCert(intermediateCA)
+	}
+
+	// Verify certificate chain, using tainted authorities as root
+	_, err := chain[0].Verify(x509.VerifyOptions{
+		Intermediates: intermediatePool,
+		Roots:         rootPool,
+	})
+
+	// TODO: may we verify if error is different to Signed by unknown authority?
```
Comment (amartinezfayo) [inline pkg/common/x509util/cert.go:97]: I think it's probably a good idea to log something if that happens.

File: pkg/agent/manager/cache/lru_cache.go [hunk 2]
```diff
@@ -521,6 +563,87 @@ func (c *LRUCache) SyncSVIDsWithSubscribers() {
...
+		c.taintedBatchProcessedCh <- struct{}{}
+	}
+}
+
+// processTaintedSVIDs identifies and removes tainted SVIDs from the cache that have been signed by the given tainted authorities.
+func (c *LRUCache) processTaintedSVIDs(entryIDs []string, taintedX509Authorities []*x509.Certificate) {
+	counter := telemetry.StartCall(c.metrics, telemetry.CacheManager, "", telemetry.ProcessTaintedSVIDs)
+	defer counter.Done(nil)
+
+	// TODO: add metric fr time
```
[the remark this answers is not in the document]
Comment (MarcosDY) [reply pkg/agent/manager/cache/lru_cache.go:623]: I already added, it was a leftover comment

File: pkg/agent/manager/cache/lru_cache.go [hunk 3]
```diff
@@ -521,6 +564,86 @@ func (c *LRUCache) SyncSVIDsWithSubscribers() {
...
+			continue
+		}
+		if isTainted {
+			taintedSVIDs++
+			delete(c.svids, entryID)
+		}
+	}
+
+	agentmetrics.AddCacheManagerTaintedSVIDsSample(c.metrics, "", float32(taintedSVIDs))
+	c.log.WithField(telemetry.TaintedSVIDs, taintedSVIDs).Debug("Tainted X.509 SVIDs")
```
Comment (amartinezfayo) [inline pkg/agent/manager/cache/lru_cache.go:644]: This should be logged at Info level, as it is in the SVID store cache.

File: pkg/agent/manager/cache/lru_cache.go [hunk 4]
```diff
@@ -521,6 +564,86 @@ func (c *LRUCache) SyncSVIDsWithSubscribers() {
...
+		// Remove processed entries from the list
+		entryIDs = entryIDs[batchSize:]
+
+		entriesLeftCount := len(entryIDs)
+		if entriesLeftCount == 0 {
+			c.log.Debug("Finished processing all tainted entries")
+			c.notifyTaintedBatchProcessed()
+			return
+		}
+		c.log.WithField(telemetry.Count, entriesLeftCount).Debug("Tainted entries left to be processed")
```
Comment (amartinezfayo) [inline pkg/agent/manager/cache/lru_cache.go:594]: I would change slightly the message to be consistent with the message about the SVIDs already tainted. I would also make this an Info level message.
```suggestion
		c.log.WithField(telemetry.Count, entriesLeftCount).Info("There are tainted X.509 SVIDs left to be processed")
```

File: pkg/agent/manager/cache/lru_cache.go [hunk 5]
```diff
@@ -521,6 +564,86 @@ func (c *LRUCache) SyncSVIDsWithSubscribers() {
...
+		processingEntries := entryIDs[:batchSize]
+
+		c.processTaintedSVIDs(processingEntries, taintedX509Authorities)
+
+		// Remove processed entries from the list
+		entryIDs = entryIDs[batchSize:]
+
+		entriesLeftCount := len(entryIDs)
+		if entriesLeftCount == 0 {
+			c.log.Debug("Finished processing all tainted entries")
```
Comment (amartinezfayo) [inline pkg/agent/manager/cache/lru_cache.go:590]: This should be logged at Info level.

File: pkg/agent/svid/rotator.go [hunk 6]
```diff
@@ -130,6 +137,43 @@ func (r *rotator) Subscribe() observer.Stream {
...
+		return nil
+	}
+
+	tainted, err := x509util.IsSignedByRoot(state.SVID, taintedAuthorities)
+	if err != nil {
+		return fmt.Errorf("failed to check if SVID is tainted: %w", err)
+	}
+
+	if tainted {
+		r.c.Log.Debug("Agent SVID is tainted by a root authority, forcing rotation")
```
Comment (amartinezfayo) [inline pkg/agent/svid/rotator.go:171]: This should be logged at Info level.

File: pkg/agent/manager/cache/lru_cache.go (hunk 3 above)
Comment (MarcosDY) [reply pkg/agent/manager/cache/lru_cache.go:644]: sorry forgot to update this.. yeah, agree that this is an user action, and we must info it

**Review by amartinezfayo — APPROVED** [review]
🎉

**Pull request merged**